CDAP-20582: Add periodic cleanup for cloned repos in appfabric

samdgupi opened the pull request:
Add a cleanup service in appfabric which deletes cloned repositories if it does not get cleaned up by `RepositoryManager`.
By default it removes directory that is created more than 2 hours ago.

Tested manually by removing the auto cleanup by `RepositoryManager` and setting low ttl.

samdgupi committed:
Add repository cleanup service in app-fabric

de-lan commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java:
```diff
@@ -70,7 +72,7 @@
  * It exposes repository CRUD apis and source control tasks that do pull/pull/list applications in linked repository.
  */
-public class SourceControlManagementService {
+public class SourceControlManagementService extends AbstractIdleService {
```
Do we need to make `SourceControlManagementService` an `AbstractIdleService`? We could make  `RepositoryCleanupService` an `AbstractIdleService` itself and start/stop in `AppFabricServer`, which makes our life easier. Example: https://github.com/cdapio/cdap/pull/15007/files

samdgupi replied:
done

de-lan commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java:
```diff
@@ -45,6 +45,7 @@
 import io.cdap.cdap.sourcecontrol.CommitMeta;
 import io.cdap.cdap.sourcecontrol.NoChangesToPullException;
 import io.cdap.cdap.sourcecontrol.NoChangesToPushException;
+import io.cdap.cdap.sourcecontrol.RepositoryCleanupService;
```
nit: this one is not needed

samdgupi replied:
done

de-lan commented on cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/guice/SourceControlModule.java:
```diff
@@ -27,5 +29,6 @@ public class SourceControlModule extends AbstractModule {
   @Override
   protected void configure() {
     bind(SourceControlOperationRunner.class).toProvider(SourceControlOperationRunnerProvider.class);
+    bind(RepositoryCleanupService.class).in(Scopes.SINGLETON);
```
Do we need the binder here?

samdgupi replied:
IMO We need the service to be a singleton to avoid multiple cleanup jobs running at the same time

de-lan commented on cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:
```diff
@@ -2056,5 +2056,8 @@ public static final class SourceControlManagement {
     public static final String GIT_COMMAND_TIMEOUT_SECONDS = "source.control.git.command.timeout.seconds";
     public static final String GIT_REPOSITORIES_CLONE_DIRECTORY_PATH =
       "source.control.git.repositories.clone.directory.path";
+    public static final String REPOSITORY_CLEANUP_INTERVAL_SECONDS = "source.control.repository.cleanup.interval" +
```
nit: just put the string to next line?

samdgupi replied:
done

de-lan commented on cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryCleanupService.java:
```diff
@@ -0,0 +1,113 @@
+
+  private void deleteExpiredRepository() {
+    try {
+    List<File> repositoryDirs = getRepositoryDirectories();
```
nit: the indent is off

de-lan commented on cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryCleanupService.java:
```diff
@@ -0,0 +1,113 @@
+    for (File repositoryDir: repositoryDirs){
+        FileTime createTimeStamp = FileUtils.getFileCreationTime(repositoryDir.toPath());
+        if (createTimeStamp != null && (currentTimeInSeconds - createTimeStamp.toMillis() / 1000) > ttlInSeconds){
+          DirUtils.deleteDirectoryContents(repositoryDir);
```
Maybe we should LOG the delete here?

samdgupi replied:
It will mostly be unactionable hence I have not added any. I can add a debug log for edge case scenarios where we need to debug issues with wrong cleanup.

de-lan reviewed (approved):
LGTM but pls address the checkstyle warnings in new code

tivv commented on cdap-common/src/main/java/io/cdap/cdap/common/utils/FileUtils.java:
```diff
@@ -114,6 +118,17 @@ public static String getNameWithoutExtension(String name) {
+   * @throws IOException if unable to read get attribute due to I/O error
+   */
+  public static FileTime getFileCreationTime(Path filePath) throws IOException {
+    return (FileTime) Files.getAttribute(filePath, "creationTime");
```
Use `Files.readAttributes(file, BasicFileAttributes.class).creationTime();` It's strong typed and less error-prone

samdgupi replied:
Updated

tivv commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -5945,5 +5945,19 @@
+    <name>source.control.repository.cleanup.interval.seconds</name>
+    <value>3600</value>
+    <description>
+      Scheduler interval for repository cleanup service, by default 1 hours
```
nit: `hour`

samdgupi replied:
Updated

tivv reviewed (commented):
Please fix checkstyle violations

samdgupi committed:
Fix checkstyle errors

samdgupi committed:
Fix additional checkstyles in constants file

samdgupi commented:
Fixed checkstyle violations in new files. Also fixed some additional checkstyle violations in constants file

samdgupi commented:
> LGTM but pls address the checkstyle warnings in new code


Fixed checkstyle violations in new files. Also fixed some additional checkstyle violations in constants file

tivv reviewed (approved):
nit: fix remaining checkstyle in Constants
